File: examples/rustyshim_client.R
rustyshim_client <- R6::R6Class(
    "rustyshim_client",
    public = list(
        initialize = function(host, username, password, request_admin, port, scheme) {
            rc <- reticulate::import("rustyshim_client")
            private$pyclient <- rc$rustyshim_connect(host, username, password, request_admin, as.integer(port), scheme)
        },
        list_actions = function() {
            private$pyclient$list_actions()
        },
        refresh_context = function() {
            private$pyclient$refresh_context()
        },
        clear_expired_items = function() {
            private$pyclient$clear_expired_items()
        },
        get_sql = function(path) {
            reader <- private$pyclient$get_sql(path)
            reader$read_all()
        }
    ),
    private = list(
        pyclient = NULL
    )
)

rustyshim_connect <- function(host, username, password, request_admin=FALSE, port=50051, scheme = "grpc+tcp") 
{
    rustyshim_client$new(host, username, password, request_admin, port, scheme)
}

File: examples/rustyshim_client.py
import pyarrow.flight as pf

class RustyShimConnection:
    class AuthHandler(pf.ClientAuthHandler):
        def __init__(self, username, password, request_admin):
            self.username = username
            self.password = password
            self.request_admin = request_admin
            self.token = None
        def authenticate(self, outgoing, incoming):
            outgoing.write(self.username)
            outgoing.write(self.password)
            outgoing.write(str(int(self.request_admin)))
            self.token = incoming.read()
        def get_token(self):
            if self.token is None:
                raise pf.FlightUnauthenticatedError("Not authenticated via SciDB")
            return self.token
    
    def __init__(self, host, username, password, request_admin, port, scheme):
        location = scheme + "://" + host + ":" + str(port)
        self.client = pf.connect(location)
        h = RustyShimConnection.AuthHandler(username, password, request_admin)
        self.client.authenticate(h)
        self.options = pf.FlightCallOptions(headers=[(b'authorization',h.get_token())])
    
    def list_actions(self):
        return self.client.list_actions(self.options)
    
    def refresh_context(self):
        return [r.body.to_pybytes().decode("utf-8") for r in self.client.do_action("REFRESH_CONTEXT", self.options)]

    def clear_expired_items(self):
        return [r.body.to_pybytes().decode("utf-8") for r in self.client.do_action("CLEAR_EXPIRED_ITEMS", self.options)]
    
    def get_sql(self, query):
        fd = pf.FlightDescriptor.for_path(query)
        fi = self.client.get_flight_info(fd, self.options)
        ep = fi.endpoints[0]
        return self.client.do_get(ep.ticket, self.options)

def rustyshim_connect(host, username, password, request_admin=False, port=50051, scheme = "grpc+tcp"):
    return RustyShimConnection(host, username, password, request_admin, port, scheme)